rust/kernel/error.rs

```diff
@@ -21,44 +21,49 @@ use core::str::{self, Utf8Error};
 ///
 /// The value is a valid `errno` (i.e. `>= -MAX_ERRNO && < 0`).
 #[derive(Clone, Copy, PartialEq, Eq)]
-pub struct Error(c_types::c_int);
+pub struct Error(NonZeroI32);
+
+/// DO NOT use this function outside of Error const initializations.
+const fn kernel_const_to_error(val: u32) -> Error {
+    Error(unsafe { NonZeroI32::new_unchecked(-(val as i32)) })
+}
 
 impl Error {
     /// Invalid argument.
-    pub const EINVAL: Self = Error(-(bindings::EINVAL as i32));
+    pub const EINVAL: Self = kernel_const_to_error(bindings::EINVAL);
```

Comment:
I am unfamiliar with Rust macros, but I wonder if it would be possible to write a macro that takes the `Error` name as the input parameter, and expands to the whole `pub const` definition? Maybe even a `static_assert!` too?

Example:
```rust
declare_const_error!(EINVAL);
// would expand to
static_assert!(bindings::EINVAL < 0 && bindings::EINVAL >= -MAX_ERRNO);
pub const EINVAL: Self = Error(unsafe { NonZeroI32::new_unchecked(-(bindings::EINVAL as i32)) });
```

Reply:
Good idea! 

But I don't know why even `static_assert!(bindings::ENOMEM < bindings::MAX_ERRNO)` doesn't work.

I got the following error:
```
error: `const` items in this context need a name
  --> rust/kernel/static_assert.rs:36:15
   |
30 | / macro_rules! static_assert {
31 | |     ($condition:expr) => {
32 | |         // Based on the latest one in `rustc`'s one before it was [removed].
33 | |         //
...  |
36 | |         const _: () = [()][!($condition) as usize];
   | |               ^ `_` is not a valid name for this `const` item
37 | |     };
38 | | }
   | |_- in this expansion of `static_assert!`
   |
  ::: rust/kernel/error.rs:42:5
   |
42 |       static_assert!(bindings::ENOMEM < bindings::MAX_ERRNO);
   |       ------------------------------------------------------- in this macro invocation
```

Reply:
If you write like what @TheSven73 suggests, `const _` in `static_assert!` will become an associated constant rather than just a normal item, so it must have a name. It'll need to expand into something like
```rust
pub const EINVAL: Self = {
	static_assert!(bindings::EINVAL < 0 && bindings::EINVAL >= -MAX_ERRNO);
	Error(unsafe { NonZeroI32::new_unchecked(-(bindings::EINVAL as i32)) })
};
```
But I think it'll be better to put the check in a const fn + const panic instead of using `static_assert` in this case:
```rust
const fn i_am_bad_at_naming_things(errno: i32) -> Error {
	assert!(errno < -(bindings::MAX_ERRNO as i32) || errno >= 0);
	Error(unsafe { NonZeroI32::new_unchecked(errno) })
}

pub const EINVAL: Self = i_am_bad_at_naming_things(-(bindings::EINVAL as i32));
```

Reply:
Got it. Thank you, Gary!!

Can I combine this? i.e. declare a `i_am_bad_at_naming_things()` function and write a macro to wrap 
`pub const EINVAL: Self = i_am_bad_at_naming_things(-(bindings::EINVAL as i32));`
into `declare_const_error!(EINVAL);`

Do you think it's a good idea??

Reply:
nvm, I opted for the first approach.

```rust
macro_rules! declare_const_error {
    ($($tt:tt)*) => {
        
        pub const $($tt)*: Self = {
            static_assert!(bindings::$($tt)* <= bindings::MAX_ERRNO && bindings::$($tt)* > 0);
            Error(unsafe { NonZeroI32::new_unchecked(-(bindings::$($tt)* as i32)) })
        };
    };
}
```

Thank you again!

Reply:
```rust
impl Error {
    /// Invalid argument.
    declare_const_error!(EINVAL);
```


```
warning: unused doc comment
  --> rust/kernel/error.rs:42:5
   |
42 |     /// Invalid argument.
   |     ^^^^^^^^^^^^^^^^^^^^^ rustdoc does not generate documentation for macro invocations
   |
   = note: `#[warn(unused_doc_comments)]` on by default
   = help: to document an item produced by a macro, the macro must produce the documentation as part of its expansion
```
Looks like this approach is not that explicit nor is it friendly for the document.

I will take the second one.

Reply:
You will have to put the doc comment inside the `declare_const_error!()` invocation. You can match on doc comments inside a macro using something like `$(#[doc = $the_doc:literal])*` I think. You will then have to place `$(#[doc = $the_doc])*` just before the `pub const` inside the macro.

Reply:
```rust
macro_rules! declare_const_error {
    ($(#[$attr:meta])* $($name:ident)*) => {
        $(#[$attr])*
		pub const $($name)*: Self = {
        	static_assert!(bindings::$($name)* <= bindings::MAX_ERRNO && bindings::$($tt)* > 0);
	        Error(unsafe { NonZeroI32::new_unchecked(-(bindings::$($name)* as i32)) })
		};
    };
}

declare_const_error! {
	/// Invalid argument.
	EINVAL
}
```
But this is more messy than just const fn perhaps.

Reply:
Agree, putting doc inside the macro invocation is bizarre to me.
I don't think it will be clearer.

Reply:
Actually, you can have everything in a single macro:
```rust
macro_rules! declare_const_error {
    ($($(#[$attr:meta])* $name:ident)*) => { $(
        $(#[$attr])*
		pub const $name: Self = {
        	static_assert!(bindings::$name <= bindings::MAX_ERRNO && bindings::$name > 0);
	        Error(unsafe { NonZeroI32::new_unchecked(-(bindings::$name as i32)) })
		};
    )* };
}

declare_const_error! {
	/// Invalid argument.
	EINVAL
	/// Out of memory.
    ENOMEM
	// and more
}
```

Reply:
I wonder if this const error declaration macro (and its invocations) should go into its own dedicated PR?
Also, perhaps it could link up with @sladyn98's #358 somehow?

Reply:
> Actually, you can have everything in a single macro:
> 
> ```rust
> macro_rules! declare_const_error {
>     ($($(#[$attr:meta])* $name:ident)*) => { $(
>         $(#[$attr])*
> 		pub const $name: Self = {
>         	static_assert!(bindings::$name <= bindings::MAX_ERRNO && bindings::$name > 0);
> 	        Error(unsafe { NonZeroI32::new_unchecked(-(bindings::$name as i32)) })
> 		};
>     )* };
> }
> 
> declare_const_error! {
> 	/// Invalid argument.
> 	EINVAL
> 	/// Out of memory.
>     ENOMEM
> 	// and more
> }
> ```

This is cool. However, I strongly disagree we use something like it.

For 
```rust
impl Error {
    pub const EINVAL: Self = kernel_const_to_error(bindings::EINVAL);
}
```
 A glimpse is enough to let you understand that we have an `EINVAL` const in the `Error` scope whose value is somehow related to the kernel's EINVAL (and you can easily infer the value of EINVAL equals to negative bindings::EINVAL).

But with
```rust
impl Error{
      declare_const_error! {
	      /// Invalid argument.
	      EINVAL
	      /// Out of memory.
          ENOMEM
	      // and more
      }
}
```

You need to spend time figuring out the magic here especially when you are not familiar with rust macro.
 It adds more mental burden to developers.

Reply:
> I wonder if this const error declaration macro (and its invocations) should go into its own dedicated PR?
> Also, perhaps it could link up with @sladyn98's #358 somehow?

Let me add a reference.